File: VB/NonPersistentObjectsDemo.Module/BusinessObjects/SerializationHelper.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.IO
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Xml.Serialization
Imports DevExpress.ExpressApp

Namespace NonPersistentObjectsDemo.Module.BusinessObjects

	Friend Module SerializationHelper
		Public Sub Load(Of T)(ByVal list As BindingList(Of T), ByVal data As String, ByVal objectSpace As IObjectSpace, ByVal acceptor As Action(Of T))
			list.RaiseListChangedEvents = False
			list.Clear()
			If data IsNot Nothing Then
				Dim serializer = New XmlSerializer(GetType(T).MakeArrayType())
				Using stream = New MemoryStream(Encoding.UTF8.GetBytes(data))
					Dim objs = TryCast(serializer.Deserialize(stream), IList(Of T))
					For Each obj In objs
						acceptor?.Invoke(obj)
						Dim tobj = objectSpace.GetObject(obj)
						Dim aobj = TryCast(tobj, IAssignable(Of T))
						If aobj IsNot Nothing Then
							aobj.Assign(obj)
						End If
						list.Add(tobj)
					Next obj
				End Using
			End If
			list.RaiseListChangedEvents = True
			list.ResetBindings()
		End Sub
		Public Function Save(Of T)(ByVal list As IList(Of T)) As String
			If list Is Nothing OrElse list.Count = 0 Then
				Return Nothing
			End If
			Dim serializer = New XmlSerializer(GetType(T).MakeArrayType())
			Using stream = New MemoryStream()
				serializer.Serialize(stream, list.ToArray())
				Return Encoding.UTF8.GetString(stream.ToArray())
			End Using
		End Function
	End Module
End Namespace

File: CS/XPO/NonPersistentDemo/NonPersistentDemo.Module/BusinessObjects/SerializationHelper.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Xml.Serialization;
using DevExpress.ExpressApp;

namespace NonPersistentObjectsDemo.Module.BusinessObjects {

    static class SerializationHelper {
        public static void Load<T>(BindingList<T> list, string data, IObjectSpace objectSpace, Action<T> acceptor) {
            list.RaiseListChangedEvents = false;
            list.Clear();
            if(data != null) {
                var serializer = new XmlSerializer(typeof(T).MakeArrayType());
                using(var stream = new MemoryStream(Encoding.UTF8.GetBytes(data))) {
                    var objs = serializer.Deserialize(stream) as IList<T>;
                    foreach(var obj in objs) {
                        acceptor?.Invoke(obj);
                        var tobj = objectSpace.GetObject(obj);
                        var aobj = tobj as IAssignable<T>;
                        if(aobj != null) {
                            aobj.Assign(obj);
                        }
                        list.Add(tobj);
                    }
                }
            }
            list.RaiseListChangedEvents = true;
            list.ResetBindings();
        }
        public static string Save<T>(IList<T> list) {
            if(list == null || list.Count == 0) {
                return null;
            }
            var serializer = new XmlSerializer(typeof(T).MakeArrayType());
            using(var stream = new MemoryStream()) {
                serializer.Serialize(stream, list.ToArray());
                return Encoding.UTF8.GetString(stream.ToArray());
            }
        }
    }
}
